### PR TITLE
refactor(cursor): remove Path B legacy code, add first-run guard

### PR DESCRIPTION
**Pull request by rangemer333-cell**
## Summary

- **Remove Path B (legacy) from cursor-hook-input.ts**: Delete the manual ~30-field allowlist path, keeping only the canonical prefix-based passthrough (Path A). Path B was missing enriched fields (`gen_ai.step.id`, `gen_ai.response.id`, `gen_ai.output.messages`) that Path A transparently carries. File reduced from 428 to ~170 lines, 15 dead functions/constants/imports removed.

- **Add first-run guard**: Override `collect()` so that when the daemon starts with no prior offset state (fresh deployment or state reset), the offset jumps to the current JSONL file end. This prevents replaying historical records that would produce duplicate traces. Subsequent polls collect only newly appended records.

- **Refactor react-assembler**: Extract `buildDeltaMessages` helper and harden the fast-path for `composer-2.5-fast` sessions. Split input messages into delta + full to avoid redundant payload.

- **Fix tests**: Adapt to `MockStateStore.set(inputId, state)` API, create fast-poll instance for first-run guard timing verification. All 13 cursor-hook-input tests pass.

## Bug fixes

| Bug | Fix |
|-----|-----|
| first-run guard 日期轮转误触发 | 条件从 `!state.lastFile \|\| state.lastFile !== logFileName` 改为 `!state.lastFile`，base class 已处理日期轮转 offset reset |
| first-run guard 文件不存在时未写 state | 文件不存在/为空时仍写 `{lastFile, lastOffset:0}` 标记 guard 已执行，防止下次 poll 重入 skip 新记录 |
| buildDeltaMessages 共享引用 | push 到 cumulativeInputMessages 时浅拷贝 `{ ...toolMessage, parts: [...toolMessage.parts] }` |
| step-opening 重复代码 | 提取 `openImplicitStep` helper 合并 composer-2.5-fast 和 buffered-tools-only 两处 |

## Breaking change note

`gen_ai.input.messages` 语义变更：从 per-step delta（仅当前 step 的增量消息）改为 cumulative full messages（截至当前 step 的完整对话上下文）。下游消费者如需按 step 区分新增消息，应改用 `gen_ai.input.messages_delta`。

## Changes

| File | Change |
|------|--------|
| `src/inputs/cursor-hook/cursor-hook-input.ts` | Remove Path B, add first-run guard with edge case fixes |
| `tests/unit/inputs/cursor-hook-input.test.ts` | Fix MockStateStore API, add date rollover + file-not-exist tests (13 tests) |
| `assets/hooks/cursor/react-assembler.mjs` | Extract buildDeltaMessages/openImplicitStep, fix shared reference |
| `tests/unit/hooks/cursor-react-assembler.test.ts` | Assembler test coverage |

## Test plan

- [x] `npx vitest run tests/unit/inputs/cursor-hook-input.test.ts` — 13 tests pass
- [x] `npx vitest run tests/unit/hooks/cursor-react-assembler.test.ts` — assembler tests pass
- [x] `npx vitest run` — full suite 1275 tests pass
- [ ] E2E: start daemon with empty state, verify no historical records replay
- [ ] E2E: start daemon before Cursor creates JSONL, verify first record collected
- [ ] E2E: run a Cursor session, verify all enriched fields present in output

### CODE REVIEW

**Review by ralf0131 — APPROVED** [review]
### Summary
Refactors cursor-hook input and react-assembler: (1) removes legacy Path B manual field-mapping in favor of the canonical prefix-based passthrough (Path A) that carries all enriched fields; (2) adds a first-run guard that skips existing JSONL history on fresh daemon start to prevent duplicate traces; (3) extracts `buildDeltaMessages` to emit both `gen_ai.input.messages_delta` (per-step delta) and `gen_ai.input.messages` (full cumulative context), with proper deep cloning via `cloneMessages`.

### Findings
No blocking issues. Code quality is high — well-documented design decisions, consistent with existing patterns.

**[Info]** `tests/unit/inputs/cursor-hook-input.test.ts` — The old test `strips token fields from canonical stop records` was removed alongside Path B tests. The `stripTokenFields()` call still exists in `transformRecord` for stop events, but its behavior is no longer explicitly unit-tested. Consider adding a lightweight test to guard this path if token stripping on stop events is important for trace deduplication.

### CI
All checks pass (Analyze, CodeQL, build-and-test 18/20/22). ✅

### Test Coverage
- First-run guard: ✅ tests both "skip existing history" and "allow new records after startup"
- Delta/full messages: ✅ tests composer-2.5-fast fast-path with buffered tools
- Backward compatibility: ✅ Path A canonical entry handles all hook events

---
*Automated review by github-manager-bot*

File: src/inputs/cursor-hook/cursor-hook-input.ts [hunk 1]
```diff
@@ -93,102 +58,48 @@ export class CursorHookInput extends BaseHookInput {
...
+   * (fresh start or offset reset), skip all existing history and only
+   * read newly appended records. This prevents replaying historical
+   * turns from the JSONL file, which would produce duplicate traces.
+   */
+  protected override async collect(): Promise<AgentActivityEntry[]> {
+    const state = this.getState();
+    const today = getTodayDateString();
+    const logFileName = `${this.logPrefix}-${today}.jsonl`;
+
+    if (!state.lastFile || state.lastFile !== logFileName) {
```
Comment (linrunqi08) [inline src/inputs/cursor-hook/cursor-hook-input.ts]: 🔴 **Bug: first-run guard 在日期切换时会跳过有效记录**

条件 `!state.lastFile || state.lastFile !== logFileName` 不仅在 daemon 首次启动（无 state）时为 true，在每天日志文件轮转时也为 true（`state.lastFile` 是昨天的文件名，`logFileName` 是今天的）。

**问题流程：**
1. Daemon 跨过午夜持续运行
2. 新的一天的 JSONL 文件中已有记录（午夜前后写入）
3. 午夜后首次 poll：guard 触发，`setState({ lastFile: 今天, lastOffset: stat.size })`
4. `super.collect()` 读到 `offset == stat.size`，返回 `[]`
5. 这些记录被永久跳过

**建议修复：** 将条件改为 `!state.lastFile`（仅在真正的首次启动时触发）。日期轮转时 base class 的 `collect()` 已经正确地将 offset 重置为 0。

```suggestion
    if (!state.lastFile) {
```

**Review by linrunqi08 — COMMENTED** [review]
## Code Review Summary

整体来看，这个 PR 的核心改动（移除 Path B、提取 `buildDeltaMessages`、新增 delta/cumulative 消息追踪）方向是正确的，代码量大幅精简。以下是评审发现的问题，按严重程度排序：

| # | 严重度 | 文件 | 问题 |
|---|--------|------|------|
| 1 | 🔴 Bug | `cursor-hook-input.ts:86` | first-run guard 在日期切换时会跳过有效记录 |
| 2 | 🟡 注意 | `react-assembler.mjs:554` | `gen_ai.input.messages` 语义从 per-step delta 变为 cumulative |
| 3 | 🟡 维护 | `react-assembler.mjs:400` | step-opening 逻辑在 3 处重复，后续修改容易遗漏 |
| 4 | 🟢 效率 | `react-assembler.mjs:577` | `cloneMessages` 在每一步对增长的累积数组做深拷贝，总代价 O(n²) |
| 5 | 🟢 架构 | `cursor-hook-input.ts:81` | first-run guard 仅在 CursorHookInput 实现，其他 BaseHookInput 子类同样面临重启回放风险 |
| 6 | 🟢 防御 | `react-assembler.mjs:588` | `buildDeltaMessages` 向 delta 和 cumulative 推入同一个 toolMessage 引用 |

详见各行内评论。

File: assets/hooks/cursor/react-assembler.mjs [hunk 2]
```diff
@@ -500,12 +539,57 @@ function buildLlmRequestWithTs(reqTs, ev, ctx, stepId, userPrompt, prevToolResul
...
+    'gen_ai.input.messages': fullMessages && fullMessages.length > 0
+      ? cloneMessages(fullMessages)
+      : undefined,
     'agent.cursor.hook_event_name': ev.hook_event,
     'agent.cursor.llm_request_time_source': timeSource,
   }, ctx.runtimeConfig);
 }
 
+/** Deep clone messages so later mutations to cumulative array don't affect emitted records. */
+function cloneMessages(messages) {
```
Comment (linrunqi08) [inline assets/hooks/cursor/react-assembler.mjs:548]: 🟡 **语义变更提醒：`gen_ai.input.messages` 从 per-step delta 变为 cumulative**

旧的 `buildLlmRequestWithTs` 将 `gen_ai.input.messages` 设为当前 step 的增量消息（user prompt 或 tool results）。新代码将其设为完整的累积会话上下文，同时新增 `gen_ai.input.messages_delta` 承载原来的 per-step 语义。

这与 `output-event-schema.md` 中 `gen_ai.input.messages` 的定义（"Full messages sent to the model"）一致，所以新语义更准确。但如果有下游消费者（dashboards、OTLP exporters）依赖旧的 per-step 语义，可能会出现 token 重复计算。

**建议：** 确认下游消费者已适配，或在 PR description / changelog 中标注此为 breaking change。

File: assets/hooks/cursor/react-assembler.mjs [hunk 3]
```diff
@@ -374,6 +395,29 @@ function buildParentSteps(events, ctx) {
     }
 
     else if (ev.hook_event === 'afterAgentResponse') {
+      // composer-2.5-fast path: no afterAgentThought, tools buffered, no step opened yet.
+      // First open s1 for the buffered tools (so afterAgentResponse can claim s2 below).
+      if (currentStepId === null && pendingToolRecords.length > 0) {
```
Comment (linrunqi08) [inline assets/hooks/cursor/react-assembler.mjs:424]: 🟡 **维护风险：step-opening 逻辑在 3 处重复**

以下三处代码都手动执行 `stepRound++` → 构造 `currentStepId` → 调用 `buildDeltaMessages` + `buildLlmRequestWithTs`：
1. `openNewStep()` (~line 358-378)
2. 此处的 composer-2.5-fast 路径 (~line 400-418)
3. 尾部的 buffered-tools-only 路径 (~line 502-518)

如果 step-opening 协议变更（例如新增字段到 llm.request），只更新其中一处而遗漏另外两处的风险较高。

**建议：** 考虑提取一个共享的 `openImplicitStep(ev, isFirst, userPrompt)` helper 来消除重复。

File: assets/hooks/cursor/react-assembler.mjs [hunk 4]
```diff
@@ -500,12 +539,57 @@ function buildLlmRequestWithTs(reqTs, ev, ctx, stepId, userPrompt, prevToolResul
...
+  if (isFirst && userPrompt) {
+    deltaMessages.push({
+      role: 'user',
+      parts: [{ type: 'text', content: userPrompt }],
+    });
+  }
+  if (previousToolResults.length > 0) {
+    const toolMessage = toolResultsToMessage(previousToolResults);
+    deltaMessages.push(toolMessage);
+    cumulativeInputMessages.push(toolMessage);
```
Comment (linrunqi08) [inline assets/hooks/cursor/react-assembler.mjs]: 🟢 **防御性建议：delta 和 cumulative 共享同一个 toolMessage 引用**

`buildDeltaMessages` 将同一个 `toolMessage` 对象同时推入 `deltaMessages`（返回值）和 `cumulativeInputMessages`（副作用）。当前 `cloneMessages()` 在记录发射时保护了正确性，但这个共享引用意味着：如果未来在 `buildDeltaMessages` 返回和 `cloneMessages` 调用之间插入了修改 `deltaMessages` 元素的代码，修改会静默传播到 cumulative 数组。

**建议：** 可以在 push 到 cumulative 时做一次浅拷贝：

```suggestion
    cumulativeInputMessages.push({ ...toolMessage, parts: [...toolMessage.parts] });
```

File: src/inputs/cursor-hook/cursor-hook-input.ts [hunk 5]
```diff
@@ -93,102 +58,48 @@ export class CursorHookInput extends BaseHookInput {
...
+    return canonicalEntry;
+  }
+
+  /**
+   * First-run guard: when the daemon starts with no prior offset state
+   * (fresh start or offset reset), skip all existing history and only
+   * read newly appended records. This prevents replaying historical
+   * turns from the JSONL file, which would produce duplicate traces.
+   */
+  protected override async collect(): Promise<AgentActivityEntry[]> {
```
Comment (linrunqi08) [inline src/inputs/cursor-hook/cursor-hook-input.ts:81]: 🟢 **架构建议：first-run guard 可以考虑下沉到 BaseHookInput**

当前 first-run guard 仅在 `CursorHookInput` 实现。但所有 `BaseHookInput` 子类（ClaudeCodeLogInput、CodexLogInput、OpenCodeLogInput、QwenCodeCliLogInput 等）都面临相同的问题：daemon 首次启动或 state 重置后，`collect()` 会从 offset 0 读取整个当天的 JSONL 文件，产生重复 trace。

如果这个 guard 逻辑是通用需要的，建议以 opt-in 选项的方式移到 `BaseHookInput` 中（例如构造函数增加 `skipHistoryOnColdStart` 参数），避免每个子类各自实现不一致的 guard 策略。

File: assets/hooks/cursor/react-assembler.mjs [hunk 6]
```diff
@@ -500,12 +539,57 @@ function buildLlmRequestWithTs(reqTs, ev, ctx, stepId, userPrompt, prevToolResul
...
+/**
+ * Build per-step delta messages and update cumulative input.
+ *
+ * - isFirst step: delta includes the user prompt (already pre-seeded in cumulative).
+ * - s2+ steps: delta includes a tool-role message from the previous step's results,
+ *   which is also appended to cumulativeInputMessages.
+ *
+ * Callers are responsible for resetting previousToolResults after this call.
+ */
+function buildDeltaMessages(isFirst, userPrompt, previousToolResults, cumulativeInputMessages) {
```
Comment (linrunqi08) [inline assets/hooks/cursor/react-assembler.mjs:571]: 🟢 **效率提示：cumulative 深拷贝在长 agentic turn 中产生 O(n²) 开销**

`cloneMessages(cumulativeInputMessages)` 在每个 step 的 `buildLlmRequestWithTs` 中被调用。累积数组每个 step 增长一条 tool-result 消息，第 k 步拷贝 k 条消息，N 步总序列化量为 O(N²)。

对于典型的 5-25 步 turn 影响不大，但如果遇到长循环（loop_count 较高、tool 输出较大的场景），可能产生可观的 CPU/GC 压力。

**可选优化：** 采用 snapshot-on-append 策略：每次向 cumulative push 时就做一次拷贝，发射时直接引用已冻结的快照，将总拷贝代价从 O(n²) 降到 O(n)。

Comment (rangemer333-cell) [reply assets/hooks/cursor/react-assembler.mjs:571]: cloneMessages 用 JSON.parse(JSON.stringify(...)) 深拷贝，每步拷贝 k 条消息，N 步总量 O(N²)。对于典型 5-25 步 turn（绝大多数场景），影响可忽略。只在 loop_count 极高且 tool output 极大时才有感知。如果后续遇到性能问题，snapshot-on-append 是合理方案。现在改是过早优化。

File: src/inputs/cursor-hook/cursor-hook-input.ts (hunk 5 above)
Comment (rangemer333-cell) [reply src/inputs/cursor-hook/cursor-hook-input.ts:81]: 目前只有 CursorHookInput 实现了 first-run guard，其他 BaseHookInput 子类（ClaudeCodeLogInput、CodexLogInput 等）确实面临同样问题。下沉到 BaseHookInput 作为 opt-in 选项（构造函数加 skipHistoryOnColdStart: boolean）是正确的架构方向。但当前只有 Cursor 被用户报告过重复 trace 问题，其他 agent 可以先不改，等有了具体 case 再统一推。

File: assets/hooks/cursor/react-assembler.mjs (hunk 4 above)
Comment (rangemer333-cell) [reply assets/hooks/cursor/react-assembler.mjs]: react-assembler.mjs 的 buildDeltaMessages 中，push 到 cumulativeInputMessages 时改为浅拷贝 { ...toolMessage, parts: [...toolMessage.parts] }，消除 delta 和 cumulative 之间的共享引用风险。

File: assets/hooks/cursor/react-assembler.mjs (hunk 3 above)
Comment (rangemer333-cell) [reply assets/hooks/cursor/react-assembler.mjs:424]: 提取 openImplicitStep helper，合并 composer-2.5-fast path 和 buffered-tools-only path 两处重复的 implicit step 开启逻辑

File: assets/hooks/cursor/react-assembler.mjs (hunk 2 above)
Comment (rangemer333-cell) [reply assets/hooks/cursor/react-assembler.mjs:548]: PR #85 description 补充了 gen_ai.input.messages 语义变更说明（delta → cumulative），提示下游消费者改用 gen_ai.input.messages_delta

File: src/inputs/cursor-hook/cursor-hook-input.ts (hunk 1 above)
Comment (rangemer333-cell) [reply src/inputs/cursor-hook/cursor-hook-input.ts]: cursor-hook-input.ts first-run guard 条件从 !state.lastFile || state.lastFile !== logFileName 改为 !state.lastFile，避免日期轮转时误跳当天的合法记录。新增日期轮转测试验证 lastFile 指向昨天文件时，今天的记录正常采集。

**Review by linrunqi08 — APPROVED** [review]
